Develop

0xxgen opened the pull request.

0xxgen committed:
- Implemented Phase Transitions
- Implemented BoosterCap

0xxgen committed:
Removed withdraw_funds functions

0xxgen committed:
Method to delete pool at the end of lifecyle

0xxgen committed:
Test utils

0xxgen committed:
Fuzzy tests + Bug fixes

0xxgen committed:
- Fuzzy Testing
- Last Bugfixes

0xxgen committed:
Test contract

0xxgen committed:
TS Examples

manolisliolios commented on packages/sources/refund.move:
```diff
@@ -0,0 +1,514 @@
+        let timeout_ts = option::borrow(&pool.timeout_ts);
+        assert!(clock::timestamp_ms(clock) >= *timeout_ts, ECurrentTimeBeforeTimeout);
+
+        if (is_funding_phase(pool)) {
+            let total_to_refund = total_to_refund(&pool.accounting);
+            let total_raised = total_raised(&pool.accounting);
+            assert!(total_raised < total_to_refund, 0);
+            
```
**The same comment as on the other PR applies here.** 

We need to move from funding phase directly into reclaim phase, not first into "claim" phase, otherwise you could theoretically bump it to "claim", do a few claims and then move into reclaim even the timeout has passed.

0xxgen replied:
I believe this is related your comment on the timeout check for `reclaim_funds` endpoints, correct? If so, it has been addressed by the newly implemented timeout check in these endpoints.

0xxgen replied:
Ok, I added an extra default time, for when setting the `timeout_ts` in the `start_funding_phase`:

`assert!(timeout_ts >= clock::timestamp_ms(clock) + MIN_FUNDING_PERIOD_MS + MIN_CLAIM_PERIOD_MS, EInvalidTimeoutTimestamp);`

That way we at least guarantee that the timeout_ts accounts for both periods. Both default periods are set to `259_200_000` corresponding to 3 days in milliseconds.

In the `start_claim_phase` I then added:

```
let current_ts = clock::timestamp_ms(clock);

// We check if there's at least 3 days difference between the clock and the timeout.
// If not, we bump the timeout_ts to correspond to a least the minimum period
if (timeout_ts - current_ts >= MIN_CLAIM_PERIOD_MS) {
    option::swap(&mut pool.timeout_ts, current_ts + MIN_CLAIM_PERIOD_MS);
};
```

This way the claim phase is guaranteed to have have the timespan of at least the minimum default time.

manolisliolios commented on packages/sources/booster.move:
```diff
@@ -0,0 +1,154 @@
+        transfer::transfer(boosted_claim_cap, affected_address);
+    }
+
+    public entry fun claim_refund_boosted(
```
Sorry for reposting this question, it's still a bit not clear to me.

In the "boosted" claims, you, as the holder of "publisher", can strictly define the "new_address" for an "affected_address", and this function is callable from that affected_address (as they'll receive the cap), but in the end of the day, all the funds will be transferred into the "new_address" that is saved. 

What is this `new_address` in that function, and why do you (`Publisher` owner) control it?

0xxgen replied:
The "boosted" claims will be allowed only if the user allocates the funds for buying and selling tokens or NFTs. We use a backend wallet associated to the user for this in our telegram trading bot, which allows users to withdraw profits, however the 150% notional remains the in the account without the user being able to transfer the funds out the backend wallet associated to the user.

The `new_address` therefore represents the user's newly generated backend wallet on the trading bot. Therefore the flow is the following:

1. We generate a backend wallet for the user and link it to the user's affected wallet by calling the permissioned endpoint `allow_boosted_claim`;
2. The user then calls `claim_refund_boosted` by signing with their affected wallet, therefore transferring the entirety of the funds to their backend wallet

The reason why `allow_boosted_claim` is permissioned is to make sure the funds are ultimately transferred to the backend wallet and the reason for the seconds step `claim_refund_boosted` requiring the affected wallet signature is to make sure the flow ultimately is consented by the user.

I hope this answers is :)

manolisliolios replied:
Thanks for elaborating!

manolisliolios commented on packages/sources/refund.move:
```diff
@@ -0,0 +1,514 @@
+    /// #### Panics
+    /// 
+    /// - If the claimer's address is not in the list of unclaimed refunds.
+    public entry fun claim_refund(
```
Do we wanna allow claiming refunds after the timeout phase has elapsed? 

If not, a check for that would also solve the problem mentioned on my other comment on this file.

0xxgen replied:
Implemented

0xxgen committed:
Added clock assertion to claim_refund endpoints

0xxgen committed:
- Set MIN_CLAIM_PERIOD_MS to 3 days
- Updated claim timeout assertion

0xxgen committed:
- Added time assertion to `start_claim_phase`
- Upgraded Sui dependency

0xxgen committed:
Added clock obj

manolisliolios commented on packages/sources/refund.move:
```diff
@@ -0,0 +1,528 @@
+    ) {
+        assert_funding_phase(pool);
+        let timeout_ts = *option::borrow(&pool.timeout_ts);
+        assert_claim_phase_time(timeout_ts, clock);
```
I think that when advancing towards claims, we should check the `MIN_CLAIM_PERIOD_MS`, like we do on the funding phase. 

Maybe split into two defaults, one for claims, one for funding periods.

0xxgen replied:
Just to confirm if I understood correctly, currently the MIN_CLAIM_PERIOD_MS check  is done when changing from:

funding phase --> claim phase

Do you suggest adding another min_time check from:

add_addresses phase --> funding phase?

0xxgen committed:
- Docstrings
- Added rinbot_address field to refund_boosted for extra safety, thus forcing the user to explicitly mention it
- `return_booster_cap` endpoint for restarting the process

0xxgen committed:
Delete function for BoostCap

0xxgen committed:
Ensured claim phase has miminum claim period

manolisliolios commented on packages/sources/refund.move:
```diff
@@ -442,6 +448,14 @@ module refund::refund {
+
+        // We check if there's at least 3 days difference between the clock and the timeout.
+        // If not, we bump the timeout_ts to correspond to a least the minimum period
+        if (timeout_ts - current_ts >= MIN_CLAIM_PERIOD_MS) {
```
Don't we want the reverse check? 
`if (timeout_ts - current_ts < MIN_CLAIM_PERIOD_MS) { ... }`?

0xxgen replied:
That's right. Changed it

0xxgen committed:
Fixed if statement in start_claim_phase

0xxgen committed:
Time check tests
